src/net/ossl.cc

```diff
@@ -1653,8 +1781,66 @@ class session : public enable_shared_from_this<session>, public session_impl {
     friend int bio_write_ex(BIO* b, const char * data, size_t dlen, size_t * written);
     friend int bio_read_ex(BIO* b, char * data, size_t dlen, size_t *readbytes);
     friend long bio_ctrl(BIO * b, int ctrl, long num, void * data);
+    friend int session_ticket_cb(SSL*, unsigned char[16], unsigned char[EVP_MAX_IV_LENGTH],
+                                 EVP_CIPHER_CTX*, EVP_MAC_CTX*, int);
 };
 
+// The following callback function is used whenever session tickets are generated or received by
+// the TLS server.  If TLS session resumption is enabled, then an AES and HMAC key are
+// generated and stored within the certificate_credentials (which is stored within the TLS session).
+// The call back uses these keys to initialize the encryption and MAC operations for both encryption (enc = 1)
+// and decryption (enc = 0).  Because the key lives with the certificate_credentials which is passed
+// to every instance of an SSL session, the same key can be used over and over again to encrypt/decrypt
+// session tickets across multiple instances of server sessions.  For more information see:
+// https://docs.openssl.org/3.0/man3/SSL_CTX_set_tlsext_ticket_key_cb/
+int session_ticket_cb(SSL * s, unsigned char key_name[16],
+                      unsigned char iv[EVP_MAX_IV_LENGTH],
+                      EVP_CIPHER_CTX * ctx, EVP_MAC_CTX *hctx, int enc) {
+    auto * sess = static_cast<const session *>(SSL_get_ex_data(s, SSL_EX_DATA_SESSION));
+    std::span<unsigned char, 16> key_name_span(key_name, 16);
+    const auto & gen_key_name = sess->_creds->get_session_ticket_keys().key_name;
+    const auto & aes_key = sess->_creds->get_session_ticket_keys().aes_key;
+    auto hmac_key_ptr = sess->_creds->get_session_ticket_keys().hmac_key.data();
+    auto hmac_key_size = sess->_creds->get_session_ticket_keys().hmac_key.size();
+    OSSL_PARAM params[3];
+    params[0] = OSSL_PARAM_construct_octet_string(OSSL_MAC_PARAM_KEY,
+                                                  const_cast<unsigned char *>(hmac_key_ptr),
+                                                  hmac_key_size);
```

Comment:
Maybe it's purely stylistic, but I prefer something like this:
```suggestion
    const auto * sess = static_cast<session *>(SSL_get_ex_data(s, SSL_EX_DATA_SESSION));
    std::span key_name_span{key_name, 16};
    std::span gen_key_name{sess->_creds->get_session_ticket_keys().key_name};
    const auto & aes_key = sess->_creds->get_session_ticket_keys().aes_key;
    std::span hmac_key{sess->_creds->get_session_ticket_keys().hmac_key};
    OSSL_PARAM params[3];
    params[0] = OSSL_PARAM_construct_octet_string(OSSL_MAC_PARAM_KEY,
                                                  const_cast<unsigned char *>(hmac_key.data()),
                                                  hmac_key.size());
```

Reply:
Or even:
```c++
    const auto * sess = static_cast<session *>(SSL_get_ex_data(s, SSL_EX_DATA_SESSION));
    auto const& session_keys = sess->_creds->get_session_ticket_keys();
    std::span key_name_span{key_name, 16};
```
and then use `session_keys.hmac_key.size()` for example.

Reply:
It seems wordy, though, so maybe:
```c++
struct session_ticket_keys {
    std::array<uint8_t, 16> name;
    std::array<uint8_t, 32> aes;
    std::array<uint8_t, 32> hmac;
```